### DISCUSSION TITLE
Debugging MSBuild - setup difficulties

### DISCUSSION THREAD

**Discussion by eturgeon123** [dotnet/msbuild#10683] [General]
I'm having difficulties setuping my environment in order to launch a simple msbuild run in the debugger. 
My understanging is that I have to set some environment variables in order to make it work since the msbuild.exe is not located in his usual location, among other things. 

Here's where I'm at right now (the last exception I got), and below I pasted my launch configuration : 

C:\Program Files\dotnet\sdk\9.0.100-rc.1.24452.12\sdks\Microsoft.NET.Sdk\Sdk\Sdk.props(49,11): 
  error MSB4226: The imp
orted project "C:\Users\admin\Documents\....\msbuild-main\artifacts\bin\MSBuild\Debug\net472\Current\Microsoft.Commo
n.props" was not found. Also, tried to find "Current\Microsoft.Common.props" in the fallback search path(s) for $(MSBui
ldExtensionsPath) - "C:\Program Files (x86)\MSBuild" . These search paths are defined in "C:\Users\admin\Documents\...\msbuild-main\artifacts\bin\MSBuild\Debug\net472\MSBuild.exe.Config".

 {
  "profiles": {
    "MSBuild": {
      "commandName": "Project",
      "commandLineArgs": "\"C:\\..........sln\"",
        "environmentVariables": {
            "MSBuildSDKsPath": "C:\\Program Files\\dotnet\\sdk\\9.0.100-rc.1.24452.12\\sdks",
            "DOTNET_ROOT": "C:\\Program Files\\dotnet",                       
            "MSBuildExtensionsPath": "C:\\Program Files\\Microsoft Visual Studio\\2022\\Community\\MSBuild\\Current"
            
        }
    }
  }
}

setting the MSBuildExtensionsPath does nothing in fact. MSBuildExtensionsPath ends up being the location of the msbuild.exe of my build (msbuild-main\artifacts\bin\MSBuild\Debug\net472). I'm not even sure it is necessary to try to set those variables.

**Comment by GangWang01** [comment]
It sounds like you want to debug msbuild.exe. If so, you can refer to [walkthroughs](https://github.com/dotnet/msbuild/blob/main/documentation/wiki/Building-Testing-and-Debugging-on-Full-Framework-MSBuild.md#walkthroughs).

**Comment by eturgeon123** [comment]
Thank you. Running the build.cmd produced the artifact folder with the msbuild in it (artifacts\bin\bootstrap\net472\MSBuild\Current\Bin\MSBuild.exe), however when I launch the debugger with the project msbuild as startup project, I get the following errors : 

 error MSB4236: The SDK 'Microsoft.NET.Sdk' specified could not be found.

This is why yesterday I was trying to set some environment variable to help it find the sdk, but I find it weird that this would not be mentionned in the wiki.
In the debugging section of the wiki, the environment variable MSBUILDDEBUGONSTART is mentionned, so I gave it a shot, but it changed nothing. 

I also noticed that when I launch the debugger, the .exe that is launched is not the one in the "Current" folder (artifacts\bin\bootstrap\net472\MSBuild\Current\Bin\MSBuild.exe). Which would explain the error I had in the first place, but I don't know how to make sense of it.

**Reply by GangWang01** [reply]
Besides MSBuild binaries, to get msbuild.exe work also needs additional dependencies (Roslyn compilers, NuGet, etc.) as the beginning of the walkthrough introduces the "bootstrap" directory. The MSBuild solution at the repository root couldn't work by itself.

And the wiki introduces debugging MSBuild by attaching the debugger through setting the environment variable `MSBUILDDEBUGONSTART`. After running build.cmd, you need to open the MSBuild solution, then from a command prompt set the environment variable and build a project/solution using `artifacts\bin\bootstrap\net472\MSBuild\Current\Bin\MSBuild.exe`. A pop-up window will appear to let you choose the debugger. The opened MSBuild solution is the right one.

**Reply by eturgeon123** [reply]
Thank you, I had not understood that I couldn't just launch the debugger in VS pressing F5, targeting the msbuild project. I needed to run the executable in command line and then attach the debugger. You made that clear to me, thank you. I have made some progress, I can attach VS now.

However, I'm unable to successfully build any solution. I tried different solution, and this is the error I get with a very simple one :

& ".\artifacts\bin\bootstrap\net472\MSBuild\Current\Bin\MSBuild.exe" "C:\....\.....sln" 


 
Unhandled Exception: System.IO.FileNotFoundException: Could not load file or assembly 'System.Threading.Tasks.Dataflow, Version=8.0.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a' 
or one of its dependencies. The system cannot find the file specified.
   at Microsoft.Build.Execution.BuildManager.Dispose(Boolean disposing)
   at Microsoft.Build.Execution.BuildManager.Dispose() in C:\TFS\P\eturgeon123\MSBuild\src\Build\BackEnd\BuildManager\BuildManager.cs:line 1178
   at Microsoft.Build.CommandLine.MSBuildApp.BuildProject(String projectFile, String[] targets, String toolsVersion, Dictionary`2 globalProperties, Dictionary`2 restoreProperties, 
ILogger[] loggers, LoggerVerbosity verbosity, DistributedLoggerRecord[] distributedLoggerRecords, Boolean needToValidateProject, String schemaFile, Int32 cpuCount, Boolean enableNodeReuse, 
TextWriter preprocessWriter, TextWriter targetsWriter, Boolean detailedSummary, ISet`1 warningsAsErrors, ISet`1 warningsNotAsErrors, ISet`1 warningsAsMessages, Boolean enableRestore, 
ProfilerLogger profilerLogger, Boolean enableProfiler, Boolean interactive, ProjectIsolationMode isolateProjects, GraphBuildOptions graphBuildOptions, Boolean lowPriority, Boolean 
question, Boolean isBuildCheckEnabled, String[] inputResultsCaches, String outputResultsCache, Boolean saveProjectResult, BuildResult& result, Boolean reportFileAccesses, String 
commandLine) in C:\TFS\P\user\MSBuild\src\MSBuild\XMake.cs:line 1694
   at Microsoft.Build.CommandLine.MSBuildApp.Execute(String commandLine) in C:\TFS\P\user\MSBuild\src\MSBuild\XMake.cs:line 1045
   at Microsoft.Build.CommandLine.MSBuildApp.Main() in C:\TFS\P\user\MSBuild\src\MSBuild\XMake.cs:line 276


Of course, this solution does build if I do it from visual studio.

**Reply by GangWang01** [reply]
I don't hit this error. What's the Visual Studio version you use? It's probably not compatible with the latest code in main branch of MSBuild. The "bootstrap" directory acquires the additional dependencies from local installed Visual Studio.

**Reply by eturgeon123** [reply]
The latest version, 17.11.4.

**Comment by eturgeon123** [comment]
I did not mentionned it, but my goal is to modify msbuild.exe in order to remove **all** the warning it emits, since its not possible to do that with configs/arguments.

**Reply by GangWang01** [reply]
What did you try to get the warnings suppressed? And which warnings do you want to suppress?

**Reply by eturgeon123** [reply]
I tried specifying the warning to exclude in the .csproj file and in the .props files, but it seems to work only for certain warnings. I don't know why. I believe MSB3277 is an example of warning that cannot be suppressed, but maybe I'm confusing numbers. If you know a way to really suppress all warnings I'd be very happy to learn how.

**Reply by KalleOlaviNiemitalo** [reply]
You could try `-noWarn:MSB3277` in an [MSBuild response file](https://learn.microsoft.com/visualstudio/msbuild/msbuild-response-files).  Although IIRC the MSBuildWarningsAsMessages property should also be able to suppress any warning, regardless of whether it's reported by the MSBuild engine or by a task.

**Reply by KalleOlaviNiemitalo** [reply]
Another option might be to define a custom logger.  That would need less work than maintaining an entire fork of MSBuild.

**Reply by GangWang01** [reply]
I'm not sure how you specified the .csproj file or .props file to exclude the warning. Can you provide a simple sample to reproduce it?

**Reply by eturgeon123** [reply]
@GangWang01 

Here's a sample of a Directory.Build.props file in located at the root of my repos. I tried different ways to write it, it seems to work only for certain warning. MSB3277,CS0162 don't work, for instance.
``
<Project>
  <PropertyGroup>
	<NoWarn>NU1605,MSB3277,3277,CS0162</NoWarn>
  </PropertyGroup>  
</Project>
``


or directly in the .csproj : 
``
<Project Sdk="Microsoft.NET.Sdk">
  <PropertyGroup>
	<NoWarn>162</NoWarn>
  </PropertyGroup>
``

**Reply by eturgeon123** [reply]
@KalleOlaviNiemitalo 

I also tried defining a custom logger, but it does not allow to remove messages, only to add some. So the real problem remains, everywhere I use msbuild (Visual Studio, powershell, etc.) I will end up seeing all the warnings.

**Reply by KalleOlaviNiemitalo** [reply]
I haven't tried with a custom logger but I imagine the `-noConsoleLogger` option would prevent the console logger from showing the warnings, and the custom logger could then decide whether to show them or not.

But that doesn't matter if `-noWarn` works.

**Reply by eturgeon123** [reply]
Ok I'm making progress

with `<MSBuildWarningsAsMessages>MSB3277</MSBuildWarningsAsMessages>,` it is true, It does remove the warning.

I'm still stuck with the CS0219, CS0162, for instance , with this props file  : 
``
<NoWarn>NU1605,MSB3277,3277,CS0162,0162,CS0219</NoWarn>
``
``
<MSBuildWarningsAsMessages>MSB3277</MSBuildWarningsAsMessages>
``

**Reply by KalleOlaviNiemitalo** [reply]
AFAICT the `MSBuildWarningsAsMessages` property requires semicolons `;` rather than commas `,`.

**Reply by KalleOlaviNiemitalo** [reply]
The .NET SDK automatically copies `NoWarn` to `MSBuildWarningsAsMessages` if `MSBuildWarningsAsMessages` is empty.
It also passes `NoWarn` to various tasks; at least some of them expect semicolons as separators.
I therefore suggest defining only `<NoWarn>NU1605;MSB3277;CS0162;CS0219</NoWarn>` (with semicolons as shown) and not defining `MSBuildWarningsAsMessages`.

**Reply by eturgeon123** [reply]
the  -nowarn switch seems to work. So I will have to retry the response file solution

**Reply by eturgeon123** [reply]
@KalleOlaviNiemitalo  

I tried it in the props file, and it does not work. I still see the MSB3277/CS0162 warnings when I build in Visual studio.
`<NoWarn>NU1605;MSB3277;CS0162;CS0219</NoWarn>	`

**Reply by KalleOlaviNiemitalo** [reply]
Do you set a NoWarn property in the project too?  If you do, it'll override what's in Directory.Build.props, unless it explicitly appends, e.g.

```XML
<PropertyGroup>
    <NoWarn>$(NoWarn);CS1591</NoWarn>
</PropertyGroup>
```

**Reply by eturgeon123** [reply]
I removed it from the csproj, but I still see the warnings.

**Reply by eturgeon123** [reply]
Worth mentionning , I just tried the /clp:ErrorsOnly switch, which could sometimes be pleasant. But with it the build time goes from 74 seconds to 155 seconds. Which makes no sense to me .
